fuel-txpool/src/service.rs

```diff
@@ -1,143 +1,137 @@
+                }));
+                return true;
+            } else {
+                warn!("Starting TxPool service that is stopping");
             }
+        } else {
+            warn!("Service TxPool is already started");
         }
-        res
+        false
     }
 
-    /// find all tx by its hash
-    async fn find(&self, hashes: &[TxId]) -> Vec<Option<TxInfo>> {
-        let mut res = Vec::with_capacity(hashes.len());
-        let pool = self.txpool.read().await;
-        for hash in hashes {
-            res.push(pool.txs().get(hash).cloned());
+    pub async fn stop(&self) -> Option<JoinHandle<()>> {
+        let mut join = self.join.lock();
+        let join_handle = join.take();
+        if let Some(join_handle) = join_handle {
+            let _ = self.sender.send(TxPoolMpsc::Stop).await;
+            let receiver = self.receiver.clone();
+            Some(tokio::spawn(async move {
+                let ret = join_handle.await;
+                *receiver.lock().await = ret.ok();
+                ()
+            }))
+        } else {
+            None
         }
-        res
     }
 
-    async fn find_one(&self, hash: &TxId) -> Option<TxInfo> {
-        self.txpool.read().await.txs().get(hash).cloned()
+    pub fn subscribe_ch(&self) -> broadcast::Receiver<TxStatusBroadcast> {
+        self.broadcast.subscribe()
     }
 
-    /// find all dependent tx and return them with requsted dependencies in one list sorted by Price.
-    async fn find_dependent(&self, hashes: &[TxId]) -> Vec<ArcTx> {
-        let mut seen = HashMap::new();
-        {
-            let pool = self.txpool.read().await;
-            for hash in hashes {
-                if let Some(tx) = pool.txs().get(hash) {
-                    pool.dependency()
-                        .find_dependent(tx.tx().clone(), &mut seen, pool.txs());
-                }
-            }
-        }
-        let mut list: Vec<ArcTx> = seen.into_iter().map(|(_, tx)| tx).collect();
-        // sort from high to low price
-        list.sort_by_key(|tx| Reverse(tx.gas_price()));
+    pub fn sender(&self) -> &mpsc::Sender<TxPoolMpsc> {
+        &self.sender
+    }
 
-        list
+    pub async fn insert(
+        &self,
+        txs: Vec<Arc<Transaction>>,
+    ) -> oneshot::Receiver<Vec<anyhow::Result<Vec<Arc<Transaction>>>>> {
+        let (response, receiver) = oneshot::channel();
+        let _ = self.sender.send(TxPoolMpsc::Insert { txs, response }).await;
+        receiver
     }
 
-    /// Iterete over `hashes` and return all hashes that we dont have.
-    async fn filter_by_negative(&self, tx_ids: &[TxId]) -> Vec<TxId> {
-        let mut res = Vec::new();
-        let pool = self.txpool.read().await;
-        for tx_id in tx_ids {
-            if pool.txs().get(tx_id).is_none() {
-                res.push(*tx_id)
-            }
-        }
-        res
+    pub async fn find(&self, ids: Vec<TxId>) -> oneshot::Receiver<Vec<Option<TxInfo>>> {
```

Comment:
consider awaiting the oneshot here to simplify the interface.

```suggestion
    pub async fn find(&self, ids: Vec<TxId>) -> Vec<Option<TxInfo>> {
```

Reply:
Ultimately it's the same behavior for the client/caller either way right? They should still use timeouts on potentially long-running async-await methods if needed, I'm having a hard time imagining how I would handle the oneshot result differently than any other regular future.

Reply:
sometimes we want to wait for a return, sometimes we don't care. TxPool return of `include` from the user side matters as it gives a reason for not including, TxPool return of `include` from p2p side can be dropped. Having more options is not a bad thing and it is addressed here: https://github.com/FuelLabs/fuel-core/compare/rakita/txpool_channels...rakita/txpool_sender
Where you can use both direct call and with `DerefMut` use `channel` `send` if needed.